### PR TITLE
Provide updateActionModify API in AutoUpdate

### PR DESCRIPTION
**Pull request by nanonaren**
Fixes yesodweb/wai#535

### CODE REVIEW

**Comment by snoyberg** [comment]
Travis builds failed due to (at least) missing imports.

**Comment by nanonaren** [comment]
How did that happen? Why would it compile for me?

**Comment by snoyberg** [comment]
You're probably using a newer version of GHC.

**Comment by nanonaren** [comment]
Right. I am using 7.10.3 and the Travis build doesn't show the missing import for `<*>` in that version but does in 7.8.4 and 7.6.3. How do I fix this?

**Comment by nanonaren** [comment]
I guess I'll just not use `<*>`. Probably the simplest thing to do.

**Comment by snoyberg** [comment]
By adding the missing import statement.

**Comment by nanonaren** [comment]
Adding the import statement raises the complaint
```
/home/narens/Documents/wai/auto-update/Control/AutoUpdate.hs:42:1: Warning:
    The import of ‘Control.Applicative’ is redundant
      except perhaps to import instances from ‘Control.Applicative’
    To import instances alone, use: import Control.Applicative()
```

Is that okay?

**Comment by snoyberg** [comment]
It's OK. The usual solution is to use CPP to make the import conditional.

**Comment by nanonaren** [comment]
Sorry, I could only find one example of `#ifdef GHC_7_4` in the code. Would this be fine
```
#ifndef GHC_7_10
import           Control.Applicative     ((<*>))
#endif
```

**Comment by snoyberg** [comment]
I'm on my phone right now so I can't find an example myself, but there
should be lots of usages of MIN_VERSION scattered around

File: auto-update/Control/AutoUpdate.hs [hunk 1]
```diff
@@ -90,6 +92,11 @@ data UpdateSettings a = UpdateSettings
     -- Default: does nothing.
     --
     -- @since 0.1.0
+    , updateActionModify   :: Maybe (a -> IO a)
+    -- ^ Optional action to be performed to get the current value
+    -- and updating it if necessary.
```
Comment (MaxGabriel) [inline auto-update/Control/AutoUpdate.hs]: This sentence switches tense—I think you want "to get the current value and update it if necessary."

File: auto-update/Control/AutoUpdate.hs [hunk 2]
```diff
@@ -90,6 +92,11 @@ data UpdateSettings a = UpdateSettings
     -- Default: does nothing.
     --
     -- @since 0.1.0
+    , updateActionModify   :: Maybe (a -> IO a)
```
Comment (MaxGabriel) [inline auto-update/Control/AutoUpdate.hs]: This function needs to be exported in the export list above, btw

**Comment by MaxGabriel** [comment]
Looks like there are some auto-update related compiler errors in the latest Travis results, btw: https://travis-ci.org/yesodweb/wai/jobs/127737868

**Comment by nanonaren** [comment]
It looks like this change cannot be made without breaking the API. So, this error
```
Network/Wai/Handler/Warp/Date.hs:31:27:

    Couldn't match type ‘()’ with ‘ByteString’

    Expected type: Control.AutoUpdate.UpdateSettings GMTDate

      Actual type: Control.AutoUpdate.UpdateSettings ()

    In the expression: defaultUpdateSettings

    In the first argument of ‘mkAutoUpdate’, namely

      ‘defaultUpdateSettings

         {updateAction = formatHTTPDate <$> getCurrentHTTPDate}’
```

happens because in the use above `updateAction` is updated to the type `IO GMTDate` but `updateActionModify` remains unchanged. It will have to be explicitly set.

**Comment by snoyberg** [comment]
That's unfortunate. I don't see a way of making this happen then.

**Comment by snoyberg** [comment]
There's another approach that could be taken I think: add a new function `mkAutoUpdateWithModify` that takes the modification value as a separate argument.

**Comment by MaxGabriel** [comment]
I think this is good to merge I think

**Comment by snoyberg** [comment]
Thanks!

**Pull request merged**